src/Annotation.cpp

```diff
@@ -0,0 +1,136 @@
+#include "Annotation.h"
+#include "Debug.h"
+#include "FunctionAnnotation.h"
+#include "ParameterAnnotation.h"
+#include "ReturnAnnotation.h"
+#include "StructAnnotation.h"
+#include "Utils.h"
+
+Annotation::Annotation() {
+  this->annotationType = AnnotationType::UndefinedAnnotation;
+}
+
+bool Annotation::annotationIsUndefined() {
+  return this->annotationType == AnnotationType::UndefinedAnnotation;
+}
+
+Annotation *generateAnnotation(const std::string &rawAnnotationString) {
+  AnnotationType annoType = AnnotationType::UndefinedAnnotation;
+  std::vector<std::string> chunks =
+      dataflow::splitString(rawAnnotationString, ' ');
+
+  FunctionAnnotation *UndefinedAnnotation =
+      new FunctionAnnotation(annoType, {}, "");
+
+  // raw annotation
+  // TOOL_CHECKER (Calls || MustCall) target = {FUNCTION ||
+  // STRUCT}(name).?PARAM(int).?FIELD(str) methods = str examples:
+  /*
+  TOOL_CHECKER Calls target = FUNCTION(does_free) methods = free
+  -- FunctionAnnotation; not anticipated that we'll be needing it but some
+  handling for it is here just in case
+
+  TOOL_CHECKER Calls target = FUNCTION(does_free).PARAM(1) methods = free
+  -- ParameterAnnotation
+
+  TOOL_CHECKER Calls target = FUNCTION(creates_obligation).PARAM(2).FIELD(x)
+  methods = free
+  -- ParameterAnnotation with field
+
+  TOOL_CHECKER MustCall target = FUNCTION(creates_obligation).RETURN methods =
+  free
+  -- ReturnAnnotation
+
+  TOOL_CHECKER Calls target = FUNCTION(does_something).RETURN.FIELD(x) methods =
+  free
+  -- ReturnAnnotation with field
+
+  TOOL_CHECKER MustCall target = STRUCT(my_struct).FIELD(x) methods = free
+  -- StructAnnotation; always has FIELD specifier
+  */
+
+  if (chunks[0] != "TOOL_CHECKER") {
+    return UndefinedAnnotation;
+  }
+
+  if (chunks[1] == "Calls") {
+    annoType = AnnotationType::CallsAnnotation;
+  } else if (chunks[1] == "MustCall") {
+    annoType = AnnotationType::MustCallAnnotation;
+  }
+
+  if (annoType == AnnotationType::UndefinedAnnotation) {
+    return UndefinedAnnotation;
+  }
+
+  std::string methodsString;
+
+  for (int i = 7; i < chunks.size(); i++) {
```

Comment:
I think the `7` here comes from the grammar given above: there are 7 required items, and then anything after that must be a method argument? If that's correct, you should add a comment to explain it (it wasn't obvious to me, and I had to figure that out by reading the code carefully).

Reply:
If I'm wrong, then you definitely need a comment, so that I can understand it :)

Reply:
Yes you'd be correct; I'll add the following example as a comment to explain each part of `chunks` 
For the raw string "TOOL_CHECKER Calls target = FUNCTION(creates_obligation).PARAM(2).FIELD(x) methods = free1, free2"
  chunks = ['TOOL_CHECKER', 'Calls', 'target', '=', 'FUNCTION(creates_obligation).PARAM(2).FIELD(x)', 'methods', '=', 'free1,', 'free2']
  chunks[0] should be TOOL_CHECKER to help verify it's one of our own annotations and not an unrelated string
  chunks[1] is annotation type (Calls/MustCall)
  chunks[2] is "target" keyword, chunks[3] is "="
  chunks[4] is the target itself (FUNCTION/STRUCT potentially with sensible PARAM and FIELD specifiers)
  chunks[5] is "methods" keyword, chunks[6] is "="
  chunks[7..end] are all the methods

Reply:
Thanks, that sounds great - I think it would make this code a lot clearer.